Create a Readme in `lib` (#244044)
NixOS/nixpkgs: lib/README.md

# Nixpkgs lib

> This folder contains the implementation, documentation and tests for the Nixpkgs `lib` library.

Reviewer: ```suggestion
This directory contains the implementation, documentation and tests for the Nixpkgs `lib` library.
```

This is completely minor, but using "directory" consistently would probably help beginners in particular.

## Overview

The evaluation entry point for `lib` is [`default.nix`](default.nix).
This file evaluates to an attribute set containing two separate kinds of attributes:
- Sub-libraries:
  Attribute sets grouping together similar functionality.
  Each sub-library is defined in a separate file usually matching its attribute name.

  Example: `lib.lists` is a sub-library containing list-related functionality such as `lib.lists.take` and `lib.lists.imap0`.
  These are defined in the file [`lists.nix`](lists.nix).

- Aliases:
  Attributes that point to an attribute of the same name in some sub-library.

  Example: `lib.take` is an alias for `lib.lists.take`.

> Most files this directory are definitions of sub-libraries, but there are a few others:

Reviewer 2: ```suggestion
Most files in this directory are definitions of sub-libraries, but there are a few others:
```

- [`minver.nix`](minver.nix): A string of the minimum version of Nix that is required to evaluate Nixpkgs.
- [`tests`](tests): Tests, see [Running tests](#running-tests)
  - [`release.nix`](tests/release.nix): A derivation aggregating all tests
  - [`misc.nix`](tests/misc.nix): Evaluation unit tests for most sub-libraries

>   - `*.sh`: Bash script that run tests for specific sub-libraries

Reviewer 2: ```suggestion
  - `*.sh`: Bash scripts that run tests for specific sub-libraries
```

>   - Rest: Support files for the tests

Reviewer 2: ```suggestion
  - All other files in this directory exist to support the tests
```

This doesn't fit the format of the previous lines, which is actually ideal in this case.

- [`systems`](systems): The `lib.systems` sub-library, structured into a directory instead of a file due to its complexity

> - [`path`](path): The `lib.path` sub-library including design document and tests

Reviewer 2: ```suggestion
- [`path`](path): The `lib.path` sub-library, which includes tests as well as a document describing the design goals of `lib.path`
```

> - Rest: Sub-libraries

Reviewer 2: ```suggestion
- All other files in this directory are sub-libraries
```

### Module system

The [module system](https://nixos.org/manual/nixpkgs/#module-system) spans multiple sub-libraries:
- [`modules.nix`](modules.nix): `lib.modules` for the core functions and anything not relating to option definitions
- [`options.nix`](options.nix): `lib.options` for anything relating to option definitions
- [`types.nix`](types.nix): `lib.types` for module system types

## Reference documentation

Reference documentation for library functions is written above each function as a multi-line comment.
These comments are processed using [nixdoc](https://github.com/nix-community/nixdoc) and [rendered in the Nixpkgs manual](https://nixos.org/manual/nixpkgs/stable/#chap-functions).

> The comment format is described in the nixdoc README [here](https://github.com/nix-community/nixdoc#comment-format).

Reviewer: ```suggestion
The nixdoc README describes the [comment format](https://github.com/nix-community/nixdoc#comment-format).
```

> See the [chapter on contributing to the manual](https://nixos.org/manual/nixpkgs/#chap-contributing) for how to build the manual.

Reviewer: ```suggestion
See the [chapter on contributing to the Nixpkgs manual](https://nixos.org/manual/nixpkgs/#chap-contributing) for how to build the manual.
```

We have to be explicit until further notice.

## Running tests

> All library tests can run by building the derivation in [`tests/release.nix`](tests/release.nix):

Reviewer 2: ```suggestion
All library tests can be run by building the derivation in [`tests/release.nix`](tests/release.nix):
```

```bash
nix-build tests/release.nix
```

> Some commands for quicker iteration over parts are also available:

Reviewer 2: ```suggestion
Some commands for quicker iteration over parts of the test suite are also available:
```

```bash

> # Run all evaluation unit tests in tests/misc.nix

Reviewer 3: ```suggestion
# Run all evaluation unit tests in tests/misc.nix
# if the resulting list is empty, all tests passed
```

See the file.

nix-instantiate --eval --strict tests/misc.nix

# Run the module system tests
tests/modules.sh

# Run the lib.sources tests
tests/sources.sh

# Run the lib.filesystem tests
tests/filesystem.sh

# Run the lib.path property tests
path/tests/prop.sh

> ```

Reviewer 3: This should be documented at least in the files themselves, but having a copy here is probably good.

Author: How about just linking to this README from those files to avoid duplicated information?

Reviewer 3: I think the loss of convenience due to the indirection outweighs the duplication, because this info doesn't change often.
